docs/topics/templates.txt

```diff
@@ -415,6 +415,22 @@ adds defaults that differ from Jinja2's for a few options:
 
   It defaults to an empty list.
 
+  Use context processors for non-trivial request-dependent values that you want
+  computed only once per template render (and not at all if no template is
+  rendered).
+
+  If a context processor's value isn't used in a majority of your templates or
+  isn't used more than once per template render, it would be more explicit to
+  put the function that you would use as a context processor in the template
+  context (for example, using ``jinja2.Environment`` as described below) and
+  then to call that function in the template::
+
+    {{ function(request) }}
+
+  Since context processors are run for every template that's rendered, think
+  carefully before putting expensive computations in them, especially if the
+  result isn't used in all or most of your templates.
+
```

Comment:
While working on that comment I wrote something about middleware but eventuallly I erased that sentence because I was afraid to give bad advice.

Middleware can replace context processors because they can compute a value that depends on the request and store that value as an attribute of the request. Furthermore that attribute can be lazy (like `request.user`).

However, if I'm going to monkey-patch the result on the request to avoid computing it more than once in a request, I'd rather write a function that caches its result in a "private" attribute (`request._something`).

So here's why I left middleware out of this discussion:

- they're even further from Jinja2 than Django's context processors, so less appropriate for calculating values to be used in templates
- they still require monkey patching the request object, so they're no better than a function that caches its result in a private attribute in this regard
- they require accessing directly the computed value on the request object, while both context processors and global functions create an abstraction that makes it easier to change their implementation
- to avoid making the computation for every request, middleware require a lazy attribute, which is a bit complicated and fragile (like every attempt to proxy access to another object transparently in Python).

I prefer writing a function that caches its result on the request, as explained above.

Now, if memory serves, that's where Carl and I disagree. I'm tolerant of caching stuff on the request like this. He's more cautious and prefers to stay away from such hacks. That's what I meant by "for purity" but we need a better wording.

So, assuming the computed value is needed in every template and must not be computed more than once, the real trade off is:

- use a global function — downside: requires monkey patching an attribute on the request to  cache the result of the computation
- use a context processor — downside: adds another level of abstraction that doesn't have a very strong reason for existing

At this point it's pretty much a matter of taste.